napari/utils/chunk/_cache.py

```diff
@@ -0,0 +1,104 @@
+"""ChunkCache stores loaded chunks.
+"""
+import logging
+from typing import Dict, Optional
+
+import numpy as np
+from cachetools import LRUCache
+
+from ._request import ChunkRequest
+
+LOGGER = logging.getLogger("ChunkLoader")
+
+# ChunkCache size as a fraction of total RAM, needs to be configurable
+# and even dynamically configurable. Keep it small for now until we
+# figure out how ChunkCache will work with the Dask cache.
+CACHE_MEM_FRACTION = 0.1
+
+# Once a ChunkRequest has been loaded its ChunkRequest.chunks attribute
+# is a ChunkArrays dict, it's some number of named arrays.
+ChunkArrays = Dict[str, np.ndarray]
+
+
+def _get_cache_size_bytes(mem_fraction: float) -> int:
+    """Return number of bytes the cache should use at most.
+
+    Parameters
+    ----------
+    mem_fraction : float
+        Size the cache to be this fraction of RAM, for example 0.5.
+
+    Returns
+    -------
+    int
+        Number of bytes the cache should limit itself to.
+    """
+    import psutil
+
+    return psutil.virtual_memory().total * mem_fraction
+
+
+def _getsizeof_chunks(chunks: ChunkArrays) -> int:
+    """We pass this to the cachetoosl LRUCache.
+
+    Parameters
+    ----------
+    chunks : ChunkArrays
+        The arrays stored in one cache entry.
+
+    Returns
+    -------
+    int
+        How many bytes these arrays take up in memory.
+    """
+    return sum(array.nbytes for array in chunks.values())
+
+
+class ChunkCache:
+    """Cache of previously loaded chunks.
+
+    Uses a cachetools LRUCache to implement a least recently used cache
+    that will grow in memory usage up to some limit, then it will free
+    the least recently entries so total usage does not exceed that limit.
+
+    TODO_ASYNC: Do we really want this class? If we can get our chunks
+    to use the Dask cache, whether they are Dask or not, then we could
+    just use the Dask cache only. Having two caches seems bad because
+    how do we manage the sizes? They can't both be 0.5 * RAM obviously.
+
```

Comment:
What are we going to do about `ChunkCache` vs. the Dask cache? Do we have to do something better before merge?

Reply:
A fantastic question! tbh I'm thinking more and more that we should unify our approach with dask. But short term... Well, I'd stop setting the dask cache implicitly on import.

Incidentally, it occurred to me that the chunk cache should be off in many cases, for example in the example of live training a neural network — every time we hit that slice we want to recompute, because the computation itself is changing. Perhaps it too should be off by default?

Reply:
One issue with the cache I realized is using the cache for multi-scale is bogus, at least if you are zoomed in. Because we are slicing into the data with some very idiosyncratic bounds, and basically no one is ever going to ask for those exact extents again. We should probably just disable the cache for multiple.

With octree/tiles/chunks caching multi-scale will work just fine. A non-small reason every uses tiles and no one does multi-scale like we are doing. So I think that's a TODO item here.

But the Dask vs. Chunk caches are also unresolved. I think no one has used async enough to even notice or care about the cache. But eventually it's a big issue and we need things 100% working how we want.